tutorial/20_recipes/013_wilcoxon_pruner.py

```diff
@@ -0,0 +1,350 @@
+            # accept the 2-opt reversal
+            cost += delta_cost
+            idxs[i + 1 : j + 1] = idxs[i + 1 : j + 1][::-1]
+            if cost < best_cost:
+                best_idxs[:] = idxs
+                best_cost = cost
+                remaining_patience = options.patience
+
+        if cost > best_cost:
+            # If the best solution is not updated for "patience" iteratoins,
+            # restart from the best solution.
+            remaining_patience -= 1
+            if remaining_patience == 0:
+                idxs[:] = best_idxs
+                cost = best_cost
+                remaining_patience = options.patience
+
+    return best_idxs
+
+
+###################################################################################################
+# We make a random dataset of TSP.
+
+
+def make_dataset(num_vertex: int, num_problem: int, seed: int = 0) -> np.ndarray:
+    rng = np.random.default_rng(seed=seed)
+    return rng.random((num_problem, num_vertex, 2))
+
+
+dataset = make_dataset(
+    num_vertex=100,
+    num_problem=50,
+)
+
+N_TRIALS = 50
+
+
+###################################################################################################
+# We set a very small number of SA iterations for demonstration purpose.
+# In practice, you should set a larger number of iterations.
+
+
+N_SA_ITER = 10000
+count = 0
+
+
+###################################################################################################
+# We counts the number of evaluation to know how many problems is pruned.
+
+
+num_evaluation = 0
+
+
+###################################################################################################
+# In this tutorial, we optimize three parameters: `T0`, `alpha`, and `patience`.
+#
+# `T0` (Initial Temperature)
+# -----------------------------
+#
+# In Simulated Annealing, the concept of "temperature" is an analogy to control the randomness
+# of the search process. The initial temperature, denoted as `T0`, sets the starting level of
+# this temperature. A higher initial temperature allows the algorithm to explore a wider range of
+# solutions and to accept worse solutions with higher probability, facilitating escape from local
+# optima. As the algorithm progresses, the temperature is gradually decreased, leading to a more
+# refined search around promising areas,
+# ultimately aiming for convergence towards an optimal solution.
+#
+# `Alpha` (Cooling Rate)
+# -----------------------------
+#
+# `Alpha` is a parameter that dictates the rate at which the temperature is decreased in each
+# iteration of the algorithm. It is typically a value between 0 and 1, and
+# it's used to multiply the current temperature to obtain the temperature for the next iteration,
+# following the formula `T = Alpha * T`. A smaller `Alpha` value results in a quicker decrease
+# in temperature, making the algorithm converge faster but potentially missing broader exploration.
```

Comment:
```suggestion
# `T0` and `alpha` defining the temperature schedule
# -----------------------------
#
# In simulated annealing, it is important to determine a good temperature scheduling, but
# there is no "silver schedule" that is good for all problems, so we must tune the schedule
# for this problem.
# This code parametrizes the temperature as a monomial function `T0 * (1 - t) ** alpha`, where 
# `t` progresses from 0 to 1. We try to optimize the two parameters `T0` and `alpha`.
```
I think we should put the explanation of temperature itself in the description of simulated annealing.
I think the explanation of the temperature should go